GitCommands/CommitMessageManager.cs

```diff
@@ -0,0 +1,124 @@
+﻿using System.IO.Abstractions;
+using System.Text;
+
+namespace GitCommands
+{
+    public interface ICommitMessageManager
+    {
+        /// <summary>
+        /// Reads/stores whether the previous commit shall be amended (if AppSettings.RememberAmendCommitState).
+        /// </summary>
+        bool AmendState { get; set; }
+
+        /// <summary>
+        /// The pathname of the file where a prepared (non-merge) commit message is stored.
+        /// </summary>
+        string CommitMessagePath { get; }
+
+        /// <summary>
+        /// Returns whether .git/MERGE_MSG exists.
+        /// </summary>
+        bool IsMergeCommit { get; }
+
+        /// <summary>
+        /// Reads/stores the prepared commit message from/in .git/MERGE_MSG if it exists or else in .git/COMMITMESSAGE.
+        /// </summary>
+        string MergeOrCommitMessage { get; set; }
+
+        /// <summary>
+        /// Deletes .git/COMMITMESSAGE and the file with the AmendState.
+        /// </summary>
+        void ResetCommitMessage();
+    }
+
+    public sealed class CommitMessageManager : ICommitMessageManager
+    {
+        private readonly string _amendSaveStatePath;
+        private readonly string _commitMessagePath;
+        private readonly string _mergeMessagePath;
+
+        private Encoding _commitEncoding;
+        private IFileSystem _fileSystem;
+
+        public CommitMessageManager(string workingDirGitDir, Encoding commitEncoding)
+            : this(workingDirGitDir, commitEncoding, new FileSystem())
+        {
+        }
+
+        private CommitMessageManager(string workingDirGitDir, Encoding commitEncoding, IFileSystem fileSystem)
+        {
+            _fileSystem = fileSystem;
+            _commitEncoding = commitEncoding;
+            _amendSaveStatePath = GetFilePath(workingDirGitDir, "GitExtensions.amend");
+            _commitMessagePath = GetFilePath(workingDirGitDir, "COMMITMESSAGE");
+            _mergeMessagePath = GetFilePath(workingDirGitDir, "MERGE_MSG");
+        }
+
+        public bool AmendState
+        {
+            get
+            {
+                bool amendState = false;
+
+                if (AppSettings.RememberAmendCommitState && _fileSystem.File.Exists(_amendSaveStatePath))
+                {
+                    bool.TryParse(_fileSystem.File.ReadAllText(_amendSaveStatePath), out amendState);
+                }
+
+                return amendState;
+            }
+            set
+            {
+                if (AppSettings.RememberAmendCommitState && value)
+                {
+                    _fileSystem.File.WriteAllText(_amendSaveStatePath, true.ToString());
+                }
+                else
+                {
+                    _fileSystem.File.Delete(_amendSaveStatePath);
```

Comment:
🤔 Would it no fail if the file doesn't exist or otherwise inaccessible?

Reply:
It does not fail if the file does not exist (idempotent function).
If the file is not accessible, it should throw an exception because this should not happen with a GE-private file and I am opposed to silently ignore real errors regardless of their severity.

Reply:
Fair point, however we need to be mindful of a UX - we shouldn't crash the app because of that.